### PR TITLE
feat: Add uhura

### PR DESCRIPTION
**Pull request by joloppo**
Made all resources (except Keyed) into uhura readers/writers. Added many tests.

### CODE REVIEW

File: demo/demo.py [hunk 1]
```diff
@@ -56,3 +56,5 @@ class Config:
 # 7      id8  Label_A      1008
 # 8      id9  Label_A      1009
 # 9     id10  Label_B      1010
+
+ParquetResource(path=path, test_path=local_test_path_without_testdir)  # <<- injectable??
```
Comment (arturk-vortexa) [inline demo/demo.py:60]: What does this comment mean?

Comment (joloppo) [reply demo/demo.py:60]: Ah, that was just me messing around. At that point I was checking if `test_path` would also be injected. (Which I then added across the board.) The serde classes then check that the incoming serde path called `file` 🙄 - is fully injected.

File: dynamicio/io/postgres.py [hunk 2]
```diff
@@ -107,6 +111,7 @@ def check_injections(self) -> None:
 
     def read(self) -> pd.DataFrame:
         """Handles Read operations for Postgres."""
+        self.check_injections()
         if not bool(self.sql_query) ^ bool(self.table_name):  # Xor
```
Comment (arturk-vortexa) [inline dynamicio/io/postgres.py:115]: Shall we put brackets like this 

`if not (bool(self.sql_query) ^ bool(self.table_name)):`

It's probably well defined, but at first glance it's confusing whether the not operates on the whole expression or the first boolean

**Review by matteo-pallini — COMMENTED** [review]
Left a few comments, but as discussed it may be worth do a bit of refactoring as well on Friday in person

File: dynamicio/io/file/csv.py [hunk 3]
```diff
@@ -33,14 +37,27 @@ def check_injections(self) -> None:
...
+
+    def validate(self, df: pd.DataFrame) -> pd.DataFrame:
+        if schema := self.pa_schema:
+            df = schema.validate(df)
+        return df
+
+    def cache_key(self):
+        if self.test_path:
+            return str(self.test_path)
+        return f"file/{self.path}"
```
[the remark this answers is not in the document]
Comment (joloppo) [reply dynamicio/io/file/csv.py:60]: For this one the cache path would come outside of this. Via uhura task_test_mode, so the default ones are 
```
    known_good_path: str = "tests/fixtures/output_known_good",
    input_path: str = "tests/fixtures/input",
```
then within that I just don't want the different resource types to step on each others toes, hence `file/...` `postgres/...`

File: dynamicio/io/file/csv.py [hunk 4]
```diff
@@ -9,22 +9,26 @@
...
 
-class CsvResource(BaseModel):
+class CsvResource(BaseModel, Readable[pd.DataFrame], Writable[pd.DataFrame]):
     """CSV Resource."""
 
     path: Path
     read_kwargs: Dict[str, Any] = {}
     write_kwargs: Dict[str, Any] = {"index": False}
     pa_schema: Optional[Type[SchemaModel]] = None
+    test_path: Optional[Path] = None
```
[the remark this answers is not in the document]
Comment (joloppo) [reply dynamicio/io/file/csv.py:25]: Yep let's do that with the refactor.

File: dynamicio/io/file/csv.py [hunk 5]
```diff
@@ -9,22 +9,26 @@
...
 from pandera import SchemaModel
 from pydantic import BaseModel  # type: ignore
+from uhura import Readable, Writable
 
 from dynamicio.inject import check_injections, inject
+from dynamicio.serde import CsvSerde
 
 
-class CsvResource(BaseModel):
+class CsvResource(BaseModel, Readable[pd.DataFrame], Writable[pd.DataFrame]):
```
[the remark this answers is not in the document]
Comment (joloppo) [reply dynamicio/io/file/csv.py:18]: Yep, I'm wary of composition. I had it in a state where it was composition at some point. And it was super hard to follow what was happening. 1 layer of inheritance seems fine to me though.
